gbuiltins/bls381/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+// This file is part of Gear.
+
+// Copyright (C) Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(not(feature = "std"))]
+extern crate alloc;
+
+#[cfg(not(feature = "std"))]
+use alloc::vec::Vec;
+
+use codec::{Decode, Encode};
+
+/// Constant defines codec index of [`Request::MultiMillerLoop`].
+pub const REQUEST_MULTI_MILLER_LOOP: u8 = 0;
+/// Constant defines codec index of [`Request::FinalExponentiation`].
+pub const REQUEST_FINAL_EXPONENTIATION: u8 = 1;
+/// Constant defines codec index of [`Request::MultiScalarMultiplicationG1`].
+pub const REQUEST_MULTI_SCALAR_MULTIPLICATION_G1: u8 = 2;
+/// Constant defines codec index of [`Request::MultiScalarMultiplicationG2`].
+pub const REQUEST_MULTI_SCALAR_MULTIPLICATION_G2: u8 = 3;
+/// Constant defines codec index of [`Request::ProjectiveMultiplicationG1`].
+pub const REQUEST_PROJECTIVE_MULTIPLICATION_G1: u8 = 4;
+/// Constant defines codec index of [`Request::ProjectiveMultiplicationG2`].
+pub const REQUEST_PROJECTIVE_MULTIPLICATION_G2: u8 = 5;
+
+/// Type that should be used to create a message to the bls12_381 builtin actor.
+#[derive(Encode, Clone, PartialEq, Eq, Debug)]
+#[codec(crate = codec)]
+pub enum Request {
+    /// Request to pairing multi Miller loop for *BLS12-381*.
+    ///
+    /// Encoded:
+    ///   - `a`: `ArkScale<Vec<G1Affine>>`.
+    ///   - `b`: `ArkScale<Vec<G2Affine>>`.
+    #[codec(index = 0)]
+    MultiMillerLoop { a: Vec<u8>, b: Vec<u8> },
```

Comment:
What is the problem accepting `ArkScale<_>` types themselves as parameters?
In any case the caller would have to create `a` and `b` as `ArkScale` instances and then encode those into bytes.
Isn't there a way to charge gas for decoding on the receiving end based on the total payload size only?

Reply:
Yep, we are able to go that way however there are some points we have to take into consideration:
1. the `sp-crypto-ec-utils` API requires to get encoded `ArkScale<_>` as byte vectors. It means that even if we would require `ArkScale<_>` in the Request enum so the client code will not encode the arguments we will have to encode them by ourselves. Moreover we would have to bench such encoding operation to properly charge the gas;
2. the other argument is that we also would have to bench SCALE-decoding of `ArkScale<_>` instance for the same purpose of gas counting because we need to know `G1/2`-items count to charge the gas accordingly;
3. another one point is that there is a bug in the ark-works libs - https://github.com/arkworks-rs/algebra/issues/707 so we have to check items count on equality.

So I decided to decode vectors by hand to mitigate all the mentioned issues. Should be quite optimal )

Reply:
Agree with @gshep
Anyway we're passing everything as bytes. Program -> builtin -> host -> builtin -> program